cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,442 @@
+kind: CiliumNetworkPolicy
+metadata:
+  name: "l3-egress-rule"
+spec:
+  endpointSelector:
+    matchLabels:
+      run: ssh-client
+  egress:
+  - toEndpoints:
+    - matchLabels:
+        run: sshd
+    auth:
+       enabled: true
+```
+
+* Semantics are:
+  * Apply the policy to pods with the label `run: ssh-client`
+  * Allow traffic to pods with the label `run: sshd`
+  * Require authentication for these connections
+  * Enable default deny
+* Initially, require `toEndpoints` field when the `auth` field is set
+  * Cannot provide mTLS with cluster-external peers
+  * Cannot provide mTLS with cluster entities
+* If there are two conflicting policies, one that requires authentication and one that does not require authentication, then authentication will be required
+* If there are two policies with conflicting authentication fields, reject the rule
+  * Log an error in the Cilium agent logs
+  * Emit a warning status to in-pod `cilium status` output
+  * Increment a metric, with tags
+* Gotchas / corner cases:
+  * Avoid requiring auth to kube-apiserver
+  * Avoid requiring auth to SPIRE
+
+
+### mTLS, SPIFFE, and SPIRE
+
+When building an mTLS solution, SPIFFE, the API for having workloads request an attested identity with a clear chain of custody, and SPIRE, the implementation of that API, are the state of the art. The SPIFFE project has spent a lot of time thinking deeply about how you can avoid common identity-sharing pitfalls like spoofing identities, getting identity details (commonly but not always X.509 TLS keypairs, but not always) to the workload security, and being able to easily rotate and revoke these identities when required.
+
+This proposal recommends that, rather than trying to build something that will take a while to handle all of the above sharp edges, we use SPIFFE and SPIRE as essential parts of the Cilium Service Mesh story, and associate the Cilium Security Identity on a 1:1 basis with a SPIFFE identity and the associated X.509 keypairs.
+
+Pros:
+* We’re using another CNCF project, which means that we’re both being good citizens, and avoiding building something ourselves.
+* We can avoid having to run down all the ways a fast, proof-of-concept implementation would create security issues in the future. There’s no such thing as temporary when it comes to APIs.
+* We effectively add a second factor to the Cilium Security Identity - the SPIFFE identity and the keypair in the SVID.
+* The certificates aren’t actually stored in the Kubernetes cluster as Secrets, so getting access to them is more difficult. For example, someone with filesystem access on the node does not automatically gain access to the certificates.
+
+Cons:
+* We pick up a dependency on SPIRE. We’ll need to be running a SPIRE server in the cluster somewhere, and a SPIRE agent on each node (the agent is required for the attestation process to work properly).
+
+#### Other options for SPIFFE identity
+
+##### Cert-manager CSI driver
```

Comment:
```suggestion
##### cert-manager CSI driver
```
cert-manager is always lower case even when breaking backwards compatibility with the English language :D